sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,208 @@
+
+        return result.getValueAsync().onErrorResume(
+            (error) -> {
+                logger.debug("cache[{}] resulted in error", key, error);
+                // Remove the failed task from the dictionary so future requests can send other calls..
+                this.remove(key);
+                return Mono.empty();
+            }
+        );
+    }
+
+    public void set(TKey key, TValue value) {
+        logger.debug("set cache[{}]={}", key, value);
+        AsyncLazyWithRefresh<TValue> updatedValue = new AsyncLazyWithRefresh<TValue>(value);
+        this.values.put(key, updatedValue);
+    }
+
+    public void remove(TKey key) {
+        values.remove(key);
+    }
+
+    /**
+     * This is AsyncLazy that has an additional Task that can
+     * be used to update the value. This allows concurrent requests
+     * to use the stale value while the refresh is occurring.
+     */
+    private class AsyncLazyWithRefresh<TValue> {
+        private final Function<TValue, Mono<TValue>> createValueFunc;
+        private final ReentrantLock valueLock = new ReentrantLock();
+        private final ReentrantLock removeFromCacheLock = new ReentrantLock();
+
+        private boolean removeFromCache = false;
+        private Mono<TValue> value;
+        private Mono<TValue> refreshInProgress;
+
+        public AsyncLazyWithRefresh(TValue value) {
+            this.createValueFunc = null;
+            this.value = Mono.just(value);
+            this.refreshInProgress = null;
+        }
+
+        public AsyncLazyWithRefresh(Mono<TValue> value) {
+            this.createValueFunc = null;
+            this.value = value;
+            this.refreshInProgress = null;
+        }
+
+        public AsyncLazyWithRefresh(Function<TValue, Mono<TValue>> taskFactory) {
+            this.createValueFunc = taskFactory;
+            this.value = null;
+            this.refreshInProgress = null;
+        }
+
+        public Mono<TValue> getValueAsync() {
+            Mono<TValue> valueMono = this.value;
+            if (valueMono != null) {
+                return valueMono;
+            }
+
+            valueLock.lock();
+            try {
+                if (this.value != null) {
+                    return this.value;
+                }
+
+                this.value = this.createValueFunc.apply(null);
+                return this.value;
+            } finally {
+                valueLock.unlock();
+            }
+        }
+
+        public synchronized Mono<TValue> createAndWaitForBackgroundRefreshTaskAsync(Function<TValue, Mono<TValue>> createRefreshFunction) {
+            Mono<TValue> valueMono = this.value;
+            AtomicReference<TValue> originalValue = new AtomicReference<>();
+
+            valueMono.flatMap(value -> {
+                originalValue.set(value);
+                return valueMono;
+            });
+
+            AtomicReference<Mono<TValue>> refreshMono = new AtomicReference<>();
+            valueLock.lock();
```

Comment:
Not sure how this will work, I've seen weird things when combining Reactor based calls with synchronized functions and locks, ending in all the threads blocked... Wed o seem to use ReentrantLock in https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java, might be worth checking that out.